src/crypto/algorithms/megolm.ts

```diff
@@ -1177,6 +1204,11 @@ export class MegolmEncryption extends EncryptionAlgorithm {
                     continue;
                 }
 
+                // Yield prior to checking each device so that we don't block
+                // updating/rendering for too long.
+                // See https://github.com/vector-im/element-web/issues/21612
+                await sleep(0);
```

Comment:
While I see how this would unblock the event loop and fix the linked issue, I also want to make sure this won't create a performance regression in `encryptMessage`. It's quite possible that there will be a lot of devices processed here, and [the HTML spec](https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#timers) claims that after 5 nested calls, any further `setTimeout` calls are clamped to a minimum delay, which could add up.

Perhaps it would be more appropriate for the function to only yield on cancelable operations?

```suggestion
                if (isCancelled !== undefined) await sleep(0);
```

Reply:
Interesting, I'd never heard about the minimum delay!